### PR TITLE
[processing] Allow field parameter to reference a MultipleLayers parameter

### PR DESCRIPTION
**Pull request by troopa81**
## Description

*FieldParameter* can reference a *VectorLayer*, so every fields available in the layer are displayed in the field combo box. 

While I was working on improving the interface between QGIS and OTB, I stumble on some OTB algorithms that take a multiple layers parameter and a field parameter to select a common field upon all this layers. This is for training purpose for instance where you select several vector layers and give a field which contains the class or the features you want to train on.

This is not possible in QGIS at the time. Would that make sense to add this kind of behavior ? 

In this proposition, to be able to select a field from different layers, ALL layer have to be vector ones and the field need to be in ALL the layers.

If we agree, I'll complete these PR:
- [x] Finish implementation: If parentParameter returns only one layer, we never goes in the dedicated QgsVectorLayer branch, and we never grab ownership. We have to deal with deleted layer, or deleted/modified fields.
- [x] add tests
- [x] Change Widget from modeler to be able to select a MultipleLayers from the Field Parameter widget

### CODE REVIEW

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 1]
```diff
@@ -4035,28 +4035,68 @@ void QgsProcessingFieldWidgetWrapper::setParentLayerWrapperValue( const QgsAbstr
...
-    std::unique_ptr< QgsMapLayer > ownedLayer( context->takeResultLayer( layer->id() ) );
-    if ( ownedLayer && ownedLayer->type() == QgsMapLayerType::VectorLayer )
-    {
-      mParentLayer.reset( qobject_cast< QgsVectorLayer * >( ownedLayer.release() ) );
-      layer = mParentLayer.get();
-    }
-    else
+    QgsVectorLayer *vlayer = qobject_cast< QgsVectorLayer * >( layers.at( 0 ) );
+    QgsFields fields = vlayer ? vlayer->fields() : QgsFields();
+    for ( auto layer : layers.mid( 1 ) )
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp:4045]:
```suggestion
    const  QList< QgsMapLayer * > remainingLayers = layers.mid( 1 );
    for ( QgsMapLayer* layer : remainingLayers )
```

**Comment by nyalldawson** [comment]
Sounds like a good idea to me, and the implementation is sound. Nice work!

**Comment by troopa81** [comment]
I completed the PR. @nyalldawson are you with the last modifications?

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 2]
```diff
@@ -4083,28 +4094,42 @@ void QgsProcessingFieldWidgetWrapper::setParentLayerWrapperValue( const QgsAbstr
     QgsProcessingFeatureSourceDefinition fromVar = qvariant_cast<QgsProcessingFeatureSourceDefinition>( value );
     value = fromVar.source;
   }
-  QgsVectorLayer *layer = QgsProcessingParameters::parameterAsVectorLayer( parentWrapper->parameterDefinition(), value, *context );
-  if ( layer && layer->isValid() )
+
+  bool valueSet = false;
+  QList< QgsMapLayer * > layers = QgsProcessingParameters::parameterAsLayerList( parentWrapper->parameterDefinition(), value, *context );
+  if ( !layers.isEmpty() )
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp:4100]: I'd prefer if the original code was left whenever layers contains only a single layer -- the original code has the benefit that it sets the layer for the field combo box, which means that the field list dynamically responds to changes in the layer structure (fields added, removed, renamed...)

Comment (troopa81) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: Sure, it's done

**Review by nyalldawson — APPROVED** [review]

**Comment by troopa81** [comment]
Thanks @nyalldawson

**Pull request merged**